### DISCUSSION TITLE
How do I control the tool-calling in Autogen?

### DISCUSSION THREAD

**Discussion by SrijitaRoy913** [microsoft/autogen#4449] [Q&A]
I am working on a multi-agent framework, and one of the agents calls a tool to fetch the schema of tables. The agent ends up suggesting tool calls multiple times simultaneously instead of handling it in one parameter. 

This is a sample of the output:
--------------------------------------------------------------------------------
first_assistant (to User):

***** Suggested tool call (call_qQoQXF1VY1MAWzfTTvSx1CKs): fetch_relevant_ddl *****
Arguments: 
{"table_list": ["products"]}
***********************************************************************************
***** Suggested tool call (call_DDDrgYA5DwUya0IDBWbcKZ5w): fetch_relevant_ddl *****
Arguments: 
{"table_list": ["orders"]}
***********************************************************************************

--------------------------------------------------------------------------------

However, it will be better if the tool-calling were to happen just once, like this:
--------------------------------------------------------------------------------
first_assistant (to User):

***** Suggested tool call (call_qQoQXF1VY1MAWzfTTvSx1CKs): fetch_relevant_ddl *****
Arguments: 
{"table_list": ["products", "orders"]}
***********************************************************************************

--------------------------------------------------------------------------------

Is there any way to make the calling deterministic other than telling it through the prompt? Are there any parameters on configuration that needs to be enabled while defining the tool?

**Comment by ekzhu** [comment]
There is currently no easy way to achieve this in v0.2. For deterministic tool calls, what's the scenario? Do you require another inference on the tool call result or just return the result of the tool call directly.

**Reply by SrijitaRoy913** [reply]
I want to return the result of the tool call directly. I wanted to make it deterministic because (correct me if I'm wrong) autogen works in async fashion, and multiple hits on the database simultaneously might slow down the query execution.

I was looking into ways to speed up the agentic workflow (Text-to-SQL) (4 agents). It takes ~40-50 seconds to answer, and same question asked repeatedly reduces the execution time to ~19-20 seconds. 
Can you suggest some ways that will speed up the workflow?

**Reply by ekzhu** [reply]
What is the reason for using 4 agents for text-to-sql? Is it possible to do it with just one agent with tools?

What does the workflow look like? Is it a group chat?

Generally, less agents, less orchestration, and smaller models can speed up.

**Reply by ekzhu** [reply]
I also suggest you take a look at the v0.4 version of the library.

**Reply by SrijitaRoy913** [reply]
I'm currently using these libraries:
autogen-agentchat==0.2.38.
autogen==0.4.1
pyautogen==0.3.1

I'm using GPT 4o-mini model.

The agents are in a Sequential Workflow and are configured in this way:
1. Agent 1: Understands the business use-case and outputs a task plan.
2. Retrieves schema. Has a tool to retrieve.
3. Generates SQL
4. Executes SQL and retries if errors. Has a tool to execute.

Can this flow be optimised?

**Reply by ekzhu** [reply]
> autogen-agentchat==0.2.38.
autogen==0.4.1
pyautogen==0.3.1

All these libraries are going to cause conflict, just keep autogen-agentchat

> Agent 1: Understands the business use-case and outputs a task plan.
Retrieves schema. Has a tool to retrieve.
Generates SQL
Executes SQL and retries if errors. Has a tool to execute.

1. Can you perform step 2 ahead of time? E.g., put all relevant schema in the context/input? 
2. Can you combine step 1, 3 and 4 as a single agent with iterative tool call?

**Reply by SrijitaRoy913** [reply]
Got it, I'll remove the other libraries...

1. That might not be possible because relevant schema is retrieved based on the user's question.
2. I can explore this, how to do iterative tool call?

Also, is there any way I can manage cache to speed up?

**Reply by ekzhu** [reply]
For 2. if you have a user proxy initiate conversation with the agent, the two-agent chat is iterative tool call.

> Also, is there any way I can manage cache to speed up?

What do you mean? The cache is turned on by default.

**Reply by SuMiaoALi** [reply]
I had met this problem. I solved it through provide better prompts to tool_caller agent.

**Reply by SrijitaRoy913** [reply]
My assistant agent has two tasks. How do I link multiple tools to a single agent?

**Reply by anshumankmr** [reply]
@SuMiaoALi  Do you have an example?

**Comment by ranga-tec** [comment]
Try this aproach https://github.com/[ranga-tec/Autogen-dynamic-functions](https://github.com/ranga-tec/Autogen-dynamic-functions)

**Reply by anshumankmr** [reply]
This link 404s what was the approoach?

**Reply by anshumankmr** [reply]
Was this the link? https://github.com/ranga-tec/Autogen-dynamic-functions/blob/master/app.py
